Patron empowerment for catalogs in EDS API

cwolfebsco opened the pull request:
### Patron Empowerment for Catalogs Loaded in EDS

VuFind offers an exceptional user experience. Some clients value the unified result list offered by EBSCO Discovery Service and would thus like to load their catalog into EDS but would still like the full Patron Empowerment Functionality (placing holds, recalls) available in VuFind via their ILS driver for the records.

This Pull Request proposed several changes, which will allow EDS to adopt `IlsAwareTrait` and have the ability to use `HoldsTrait`.

### A number of difficulties need to be overcome:

1. not every EDS Record should trigger a holdings call on results list and/or show a Holdings Tab in detailed record
2. EDS API unique IDs are a combination of database code and accession numbers, but ILS backends only needs the record ID (Accession Number) - also see 3
3. Catalog Accession Numbers consist of a prefix and the record id, in case of OAI-PMH harvesting into EDS, these record IDs will contain a lengthy prefix, and hyphens will have been replaced by dots. In the case of Folio that breaks the UUID

For VuFind users with their catalog in EDS API, these settings will ensure full fidelity when it comes to displaying availability and placing holds.

### Configuration Requirements

**1. ILS Driver, e.g. Folio.ini**
- configure appropriately

**2. Config.ini**
- add: [Catalog] > ilsBackends[] = EDS
- set: [Item_Status] > show_full_status = true		(optional)
- set: [Authentication] > method = ILS				(optional)

**3. EDS.ini**
- set: [Catalog] > EDSHasCatalog = true
- set: [Catalog] > CatalogDatabaseId = "cat12345a"  (use correct Catalog ID)
- set: [Catalog] > CatalogANRegex[] && CatalogANReplace[]

- Example:
CatalogANRegex[] = "/^demo\.oai\.edge\.demo\.folio\.ebsco\.com\.fs12345678\./i"
CatalogANRegex[] = "/\./"
CatalogANReplace[] = ""
CatalogANReplace[] = "-"
In this case the first CatalogANRegex removes the prefix, the second will restore hyphens for the dots, reconstructing the UUID

**4. RecordTabs.ini**
- add: [VuFind\RecordDriver\EDS] > tabs[Holdings] = HoldingsILS

### To Do:

- [x] work on tests
- [x] request community feedback to improve approach

cwolfebsco committed:
ignore DS_Store files

pesky macOS files that are super annoying

cwolfebsco committed:
add Catalog section to EDS.ini

Ability to set EDSHasCatalog
CatalogDatabaseId
RegexEx to clean Accession Numbers to arrive at actual ILS Record Number

cwolfebsco committed:
basic logic

cwolfebsco committed:
conditionally add ID & Source

this is useful to avoid calls being made for Articles or items that are not from the catalog.

cwolfebsco committed:
update holdAction to support EDS API Holds

This has a lot of room for improvement and I would very much desire feedback from the community on a better approach.

cwolfebsco committed:
Avoid showing holdings tab for EDS records...

If a record originates from EDS & it is not a catalog record, then skip the holdings tab, as it is going to fail anyway...

cwolfebsco committed:
correct logic

in RecordsTab.ini one configures holdings tab to show for EDS results.

This add on, ensure that the holdingsTab is skipped for EDS records, that are not from a catalog database

cwolfebsco committed:
attempting to fix my many whitespace errors :)

cwolfebsco committed:
fixing whitespaces (tabs vs spaces, etc)

cwolfebsco committed:
double quote vs single quote, whitespace

demiankatz commented on .gitignore:
If this is an intentional addition, maybe it's worth submitting as a separate PR, since it's a general change unrelated to the rest of this PR.

demiankatz commented on config/vufind/EDS.ini:
File should end with a line break:

```suggestion
; CatalogANReplace[] = "-"

```

demiankatz commented on config/vufind/EDS.ini:
A little extra clarification might be worthwhile:

```suggestion
; set EDSHasCatalog to true and ensure your ILS Driver has been configured in config.ini's [Catalog] section.
```

demiankatz commented on config/vufind/EDS.ini:
Maybe this would be more readable if you paired up the matching parts:

```suggestion
; CatalogANRegex[] = "/^abc\./i"
; CatalogANReplace[] = ""
; CatalogANRegex[] = "/\./"
; CatalogANReplace[] = "-"
```

demiankatz commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
Minor style suggestions:

```suggestion
        // Holds on EDS API records require a different ID.
        // This id can be obtained from the getRtacIdentifier method
```

demiankatz commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
I believe you could simplify this to:

```suggestion
        $originalId = $driver->getUniqueID();
        $id = $driver->tryMethod('getRtacIdentifier', default: $originalId);
```

demiankatz commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
You could replace this whole array with `compact('id', 'patron')` now that the variable name and key names match up.

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
It's a little unusual to change the values of these class properties as a side effect of calling a get method. I'd suggest either creating separate setters (if it makes sense to explicitly set these properties as part of the call) or else passing the values to support methods through parameters instead of storing them as a state in the class.

demiankatz commented on themes/bootstrap5/templates/record/view.phtml:
The `isVisible` method of the tab class should contain this type of logic -- that is designed to encapsulate behavior in the tab-specific code instead of having special cases in more high-level code. Is there a reason you customized the template instead?

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
The indentation here seems a bit off.

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
This seems excessively indented (looks like a tab character was used instead of two spaces):

```suggestion
  <input type="hidden" value="<?=$this->escapeHtmlAttr($this->driver->getRtacIdentifier())?>" class="hiddenId">
  <input type="hidden" value="<?=$this->escapeHtmlAttr($this->driver->getSourceIdentifier())?>" class="hiddenSource">
```

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/EDS.php:
I'm surprised PHP lets you alias this method name to itself -- if the goal is to "get it out of the way" I wonder if it's better to give it a different name:

```suggestion
        Feature\IlsAwareTrait::getRealTimeHoldings as getRealTimeHoldingsTraitVersion;
```

...though ideally it might be better still to make the trait itself more flexible so we don't need to override the method here but could instead set some properties (or something like that).

demiankatz reviewed (changes requested):
Thanks, @cwolfebsco, this is off to a great start! See below for a variety of minor suggestions.

cwolfebsco commented on .gitignore:
My bad on this one, I solved it by setting up a global git ignore on my new machine, I'll pull this change out

cwolfebsco committed:
Revert "ignore DS_Store files"

This reverts commit 74a2a6ff8bfc3e7c1e2bb6387128eecc630630f9.

cwolfebsco committed:
Update config/vufind/EDS.ini

whitespace at end of .ini required

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Update config/vufind/EDS.ini

Much nicer, thank you!

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Grouping RegEx with Replace for greater clarity

Following a suggestion by Demian Katz, which makes perfect sense

cwolfebsco committed:
Merge branch 'Patron-Empowerment-for-catalogs-in-EDS-API' of https://github.com/cwolfebsco/vufind into Patron-Empowerment-for-catalogs-in-EDS-API

cwolfebsco committed:
only one white space, sorry ...

cwolfebsco commented on config/vufind/EDS.ini:
Now that you point it out it is obviously way better :) Sorry I didn't think about that myself

cwolfebsco committed:
Update module/VuFind/src/VuFind/Controller/HoldsTrait.php

Style improvement - sorry for being so sloppy

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Merge branch 'Patron-Empowerment-for-catalogs-in-EDS-API' of https://github.com/cwolfebsco/vufind into Patron-Empowerment-for-catalogs-in-EDS-API

cwolfebsco committed:
Update themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml

indentation was off... I need to verify my VSCode settings - new install...

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Merge branch 'Patron-Empowerment-for-catalogs-in-EDS-API' of https://github.com/cwolfebsco/vufind into Patron-Empowerment-for-catalogs-in-EDS-API

cwolfebsco committed:
remove logic from template, move to HoldingsILS.php

This removes logic from the template and moves it to RecordTab/HoldingsILS.php

cwolfebsco committed:
remove whitespace at end of line 104

cwolfebsco committed:
remove unused control parantheses

cwolfebsco committed:
condense logic for setting id

(based on recommendation by Demian)

cwolfebsco commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
much cleaner indeed, I've applied this suggestion

cwolfebsco committed:
use compact as array key and var are the same

cwolfebsco committed:
"fix" identation (hopefully)

I also moved the comment into a php comment so it is invisible to the curious user, who might inspet the page

cwolfebsco commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
I'll make an update and will hope for the best :) indentation in phtml files and I are not best friends (yet)

cwolfebsco committed:
change to simple use trait

To be fair, I had misunderstood what I am doing here (yup, never used traits before VuFind). As I am trying to simply override the that getRealTimeHoldings() function, as for EDS we need a little change there of getRtacIdentifier() instead of getUniqueID

cwolfebsco commented on module/VuFind/src/VuFind/RecordDriver/EDS.php:
This was down to me not correctly understanding the traits feature of PHP - I appreciate that making the trait more flexible might be the way to go in the future, but for now I might simple change this to `use Feature\IlsAwareTrait`

cwolfebsco commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
Thank you for this suggestion, you make a great point and I have attempted to implement an improved flow here, using two setter functions

cwolfebsco committed:
use setters to configure Hold Class

based on a suggestion by Demian, and he was totally right, as it makes the code easier to read.

restructured getRealTimeHoldings() for increased clarity (though it is way more verbose now)

cwolfebsco committed:
fix comments and formatting of comment

cwolfebsco committed:
whitespace clean up

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
It makes me a little nervous that we spit back the original ID regardless of what $details['id'] contains. I'm worried there might be some scenario where the logic class gets "stuck" on an ID -- particularly if we're sharing the object between multiple backends. Is there a predictable correlation between the original ID and $details['id']? I'm wondering if we could register original IDs into a map, and then look up IDs in the map to retrieve their original versions. The map would also likely need to account for the search backend to avoid ID collisions across backends. Maybe I'm overthinking this, but I remain uncomfortable with setting a global state just to influence the outcome of a single call.

I think overall it may be a safer solution to add another argument to getHoldings() that flows through the support methods... but that's a lot of "plumbing." I'd have to spend significantly more time analyzing this to make a better recommendation, but if it would help, I can try to find some time!

demiankatz commented on module/VuFind/src/VuFind/RecordTab/HoldingsILS.php:
The comment and the code seem to contradict each other. Should this be `!$this->driver->pubTypeExcludedFromRtac()`?

demiankatz reviewed (changes requested):
One more question, in addition to my earlier replies to the existing open threads.

cwolfebsco commented on module/VuFind/src/VuFind/RecordTab/HoldingsILS.php:
The confusion is likely caused by me having used a misleading function name, as the function returns true/false in a counter intuitive way. I'll try to find a better name and push that, but I will also need to look at your bigger concern (the other comment you kindly left) and think it through at my end.

cwolfebsco committed:
Fixing a terrible function name

previous function name "pubTypeRtacEnabled" would return false when the pubTypeId == ebook, which was the type we wanted to exclude from RTAC (real time availability call).

New function Name is pubTypeRtacEnabled suggesting that you get a true if RTAC should show and a false if it should not.

cwolfebsco committed:
fix function call

cwolfebsco commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
@demiankatz There is no urgency here, but I would be most appreciative of any recommendations you can make. I think I understand your concern here, but I am not skilled enough to address it properly without more guidance.

There is a direct relationship between the unique ID of a record and its "RTAC-ID".

UniqueID from an EDS API Call is something like:
cat012345a,demo.oai.edge.demo.folio.abcdef.com.fs00001178.976cbaf6.fb02.48a2.8f82.a19203769b52

The Record ID used for placing holds is like this:
976cbaf6-fb02-48a2-8f82-a19203769b52

The whole problem comes about, as I have only a very basic understanding of how all these many parts play together :(

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
I'll try to find time to take a look at this in the next few weeks; it's been a bit hectic, but hopefully it will calm down sooner or later long enough for me to do some tinkering and make a proposal. :-)

cwolfebsco commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
@demiankatz My apologizes for impatiently following up with you - would it help to try and setup a call to talk this through? I am at a conference this week, but could make time Friday or almost anytime next week.

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
@cwolfebsco, this keeps approaching the top of my to-do list but getting pushed down by other things -- this summer has been exceptionally busy, both within the VuFind community (preparing the 10.2 release -- now fortunately done -- and finalizing architectural changes for 11.0) and outside (I'm currently short-staffed and chairing a hiring committee, which is taking up a lot of my time). I'll keep trying to get to this; maybe I can accomplish something by the end of the week. If that doesn't work, ping me on Slack next week if you think a conversation would help... or else if you can wait another few weeks, my schedule is likely to calm down once a few more milestones are met.

cwolfebsco commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
@demiankatz I totally understand and I didn't mean to rush you. In the big scheme of things this isn't all that important, of course.

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into Patron-Empowerment-for-catalogs-in-EDS-API

demiankatz committed:
Refactor link overrides.

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
I found a little bit of time this afternoon, so I did some refactoring and pushed up my changes -- see commit 7acb1d284028bb6934102107ce9b0897401da52c and let me know what you think. I'm not in a position to actually test this, so I'm not sure if it still works. Please give it a try and let me know if you need help fixing anything!

cwolfebsco commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
@demiankatz  thank you so much. I tested this on a VM I have and it works a charm. Let me know, if you would like me to provide you with API details, which has a custom catalog database in EDS - simply ping me on Slack for that.

cwolfebsco committed:
Fixtures: Patron Empowerment Tests

Fixtures with catalog record from EDS API

cwolfebsco committed:
Tests for Patron Empowerment

testing:
getRtacIdentifier()
hasCatalog()
pubTypeRtacEnabled()

cwolfebsco committed:
test isVisible

I fear that this might be entirely too complicated to test this

cwolfebsco committed:
whitespace fixes

cwolfebsco committed:
result of vendor/bin/phing php-cs-fixer

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
Great news, @cwolfebsco! I see you've continued making adjustments here. Please let me know when you are ready for me to give it another review, and I'll try to take a look when time permits. (Still a bit swamped with other things, but happy to at least read through the code again here when you're ready!). Also good to know that I can get credentials for hands-on testing when the time comes for that!

cwolfebsco commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
Yes, @demiankatz  - I have worked on some tests to make sure these new functions are tested (the best I could). I am still missing some tests - but creating those tests was not obvious :( I'll be sure to ping you, likely after Independence Day Weekend.

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
Sounds good! And if there are areas you're having particular trouble testing, feel free to let me know and I can offer suggestions. Some things are just plain hard to test, but I'm happy to give it a look in case I have an idea!

cwolfebsco commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
@demiankatz I am stuck at trying to figure out how to test module/VuFind/src/VuFind/ILS/Logic/Holds.php in particular getHoldings, generateHoldings, getRequestDetails

I am not sure how to go about it and maybe you could let me know your thoughts?

If it could stay untested, I guess we could open this up for wider review...

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
@cwolfebsco, it should be possible to add tests for those classes, though it will require significant mocking. Since we have no existing unit tests covering them, I have opened #4473 to begin building tests, so there's a foundation that can be expanded upon here. I've only started with easy "low-hanging fruit" since my time is limited this week, but I hope this can still serve as a useful example. Looking at the code for testing purposes also revealed some simplifications and modernizations that could be easily applied, so I went ahead and did those things too.

Let me know if this is any help, and how you'd like to proceed. If you need more detailed examples of testing specific methods, let me know which ones to focus on, and I can try to build upon #4473. On the other hand, if I've given you enough to run with, please feel free to do whatever you have time for.

In any case, I'm hoping that #4473 can get approved and merged fairly easily. Once that is done, it's just a question of what work should be done here, and what would be better handled in separate PRs to keep unrelated concerns in separate commits.

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
Pretty sure we don't need these any more; I think I just forgot to delete them when I did my refactoring recently...

```suggestion
```

I'm out of time for testing today, so I'm just leaving this as a suggestion -- but if you agree, please accept the suggestion and give this another quick test to be sure we're not missing something.

demiankatz reviewed (changes requested):
See below for something I just noticed while comparing the work here against the refactoring I just did in my new PR...

demiankatz committed:
Merge branch 'dev' into Patron-Empowerment-for-catalogs-in-EDS-API

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
I just merged #4473 and cleaned these out as part of the conflict resolution with that PR. Hopefully I didn't break anything, but if you can double-check, I'd be grateful! If all went well, you now have some foundations for logic tests as well. As noted earlier, let me know if I can be of more help with any of this!

cwolfebsco commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
I can confirm that these changes work correctly in my environment. Thank you so much for doing this and providing a foundation on which I hopefully can build...

cwolfebsco committed:
default back to DEFAULT_SEARCH_BACKEND

Before I started to tinker with this function this function returned:

return [
            'action' => $action, 'record' => $details['id'],
            'source' => $details['source'] ?? DEFAULT_SEARCH_BACKEND,
            'query' => $queryString, 'anchor' => '#tabnav',
        ];

Re-introducing the fallback to DEFAULT_SEARCH_BACKEND in case details['source'] is not set

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
My pleasure; thanks for prompting some long-overdue code modernization and test development. Let me know how it goes with your testing, and if you need more work from my end to clarify anything further, I can try to make time!

cwolfebsco committed:
Test formatHoldings

using some Folio raw data to ensure the formatHoldings responds as expected

cwolfebsco committed:
Test: generateHoldings

cwolfebsco committed:
fixtures: sanitize fixtures

removing specific ids and replace with simplistic examples

cwolfebsco committed:
test: getRequestDetails

testing with and without override

cwolfebsco committed:
Test: getHoldingsGroupKey

cwolfebsco committed:
Test: getHoldings (multiple "modes")

cwolfebsco committed:
Style Fixes: Ensuring consistent formatting

Updating some mistakes in consistency in this file

cwolfebsco committed:
simplified test values

replaced long UUID with easer to check strings like item_id_1, holdings_id_1 and sample_hash_key

updated fixture as well

cwolfebsco committed:
vendor/bin/phing php-cs-fixer changes

cwolfebsco committed:
set driver & ilsBackends

demiankatz commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
Maybe it would make sense to use a more generic method name here, like `getUniqueIDOverrideForRequest`, so we can implement the same method name in other drivers if we have similar needs on other platforms. You can make this new method a wrapper around `getRtacIdentifier` if you find value in keeping that name as well, but it then keeps higher-level code more general.

demiankatz commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
If you agree with the more general approach, we could also make this comment more general.

demiankatz commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
Another approach could be to add a `getUniqueIDForRequest` method in the `AbstractBase` class that's a wrapper around `getUniqueID` for default behavior, and then override it in the EDS driver. This might make the code/logic more straightforward, though it does increase the overall size of the record driver interface.

demiankatz commented on module/VuFind/src/VuFind/ILS/Logic/Holds.php:
Not sure if these parentheses are mine or yours, but in any case, I'm pretty sure we don't need them.

```suggestion
            'source' => $linkOverrides['source'] ?? $details['source'] ?? DEFAULT_SEARCH_BACKEND,
```

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/EDS.php:
We typically use short-form types in comments.

```suggestion
     * @return bool
```

demiankatz commented on module/VuFind/src/VuFind/RecordDriver/EDS.php:
```suggestion
     * @return bool
```

demiankatz commented on module/VuFind/src/VuFind/RecordTab/HoldingsILS.php:
I wonder if we could eliminate the EDS-specific check here by coming up with a more generic method name like `supportsHoldingsTab`. Then we could do something like `return $this->driver->tryMethod('supportsHoldingsTab', true);` so that the tab displays by default unless the record driver says it's not compatible, and you could move this logic into the EDS driver itself for more flexibility/encapsulation.

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
I think you could use the `callMethod` convenience method in [ReflectionTrait](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFindTest/Feature/ReflectionTrait.php) to eliminate the need for local reflection logic inside the test.

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
Is there a reason you can't just use (and/or partially mock) the existing [AvailabilityStatus](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php) class? I'm not sure exactly why an anonymous class is needed here.

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
I see there are several more direct uses of ReflectionClass here that might be able to be simplified with the ReflectionTrait. If the trait has any limitations, let me know -- might be worth improving it as needed to keep the logic separate.

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
I think the OpenURL logic was intentionally excluded from EDS because the link resolver link tends to come through the EDS API directly, which could lead to weird redundancy... but maybe I'm misremembering. Just curious if you've already looked into this detail.

demiankatz reviewed (changes requested):
@cwolfebsco, I see that this is still in draft mode, so maybe my review is premature, but since there's been activity lately, I thought I'd take another look! See below for a few new suggestions based on reading this with relatively fresh eyes. :-)

cwolfebsco committed:
Update module/VuFind/src/VuFind/ILS/Logic/Holds.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
I am absolutely convinced they were mine...

cwolfebsco committed:
Update module/VuFind/src/VuFind/RecordDriver/EDS.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Update module/VuFind/src/VuFind/RecordDriver/EDS.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
I agree, giving it a more generic name seems a meaningful enhancement.

cwolfebsco committed:
getRtacIdentifier > getUniqueIDOverrideForRequest change

While longer getUniqueIDOverrideForRequest has the benefit of being more generic

cwolfebsco commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
OpenURL should indeed be coming through based on EDS API settings - so it looks like I simply "borrowed" too much from another template. I am removing it now

cwolfebsco committed:
remove openUrl and openURLActive

openURL was omitted from EDS on purpose, removing it here

cwolfebsco commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
The truth is, that my experience with Unit Tests in PHP is sadly very limited and I asked AI for help - sorry - hanging head in shame :(

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
Please let me know if there's anything I can help clear up for you! Happy to have a chat on Slack (or here, or anywhere else convenient) if that would be useful.

demiankatz commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
```suggestion
        // Holds on API records (as opposed to Solr records; e.g. EDS) may require a different ID.
```

demiankatz commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
```suggestion
        // the gatheredDetails['id'] is the original ID, but for API Holds (e.g. EDS)
        // we may need to use the originalId. So only in that case we will set it to the
```

demiankatz reviewed (changes requested):
See below for a couple of minor suggestions to make comments more generic.

It looks like you're still working through my earlier feedback, so I haven't given anything else a closer review yet.

cwolfebsco committed:
Update module/VuFind/src/VuFind/Controller/HoldsTrait.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Update module/VuFind/src/VuFind/Controller/HoldsTrait.php

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Merge remote-tracking branch 'upstream/dev' into Patron-Empowerment-for-catalogs-in-EDS-API

cwolfebsco committed:
Merge branch 'Patron-Empowerment-for-catalogs-in-EDS-API' of https://github.com/cwolfebsco/vufind into Patron-Empowerment-for-catalogs-in-EDS-API

cwolfebsco committed:
Improve tests, by using reflection Trait

Removes repeated boiler plate for ReflectionClass

cwolfebsco commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
Good point, I tried to address it with the latest commit

cwolfebsco commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
This too should be addressed with recent commit using ReflectionTrait

cwolfebsco committed:
remove anonymous class, you se AvailabilityStatusInterface

cwolfebsco commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
I've made changes to this and provided a commit that hooks straight into \VuFind\ILS\Logic\AvailabilityStatus - let's hope this makes it better.

cwolfebsco committed:
fix whitespaces

one of these days I'll learn...

cwolfebsco committed:
isVisible now based on method supportsHoldingsTab

cwolfebsco committed:
provide supportsHoldingsTab in EDS driver

cwolfebsco committed:
adjusting tests to test isVisible

cwolfebsco committed:
fix Only arrays and Traversables can be unpacked error

cwolfebsco committed:
fix whitespaces

cwolfebsco commented on module/VuFind/src/VuFind/RecordTab/HoldingsILS.php:
I followed your guidance and moved this to the driver and agree, this makes it cleaner and less dependent on it being an EDS "feature", which I like

cwolfebsco committed:
Merge remote-tracking branch 'upstream/dev' into Patron-Empowerment-for-catalogs-in-EDS-API

cwolfebsco commented:
@demiankatz I am going to be daring and mark this as ready for review - thank you so much for all the review and guidance you already provided.

demiankatz commented on module/VuFind/src/VuFind/Controller/HoldsTrait.php:
Is this a typo? Should it say:

```suggestion
        // we may need to use the override ID. So only in that case we will set it to the
```

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
This seems like a "TODO" comment. Do you still need/want to do this, or should you remove this comment?

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
Does this simplification work? I think it should as long as the holdings data doesn't have unexpected array keys.

```suggestion
        foreach ($result['holdings'] as $i => &$holding) {
            // Add 'availability' field using the real class
            $holding['availability'] = $this->createAvailabilityStatus(
                $availabilityStatus[$i][0],
                $availabilityStatus[$i][1]
            );
        }
```

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/RecordTab/HoldingsILSTest.php:
I don't think this test really serves a purpose, since its code is identical to that of testIsVisibleTrue. I'd suggest simply removing it; I think if we wanted to test the defaulting behavior, we might need to mock things differently... but I don't think it's necessary to put more effort into it, since the true and false cases are both covered appropriately, and the expectations of the mock call already accounts for the default value being set by the code.

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
This seems potentially dangerous -- I believe the hiddenId and hiddenSource are used for multiple functions (embedded records, AJAX save statuses, etc.). Unless I'm misunderstanding, this code seems to remove the values for all records except catalog records, and it changes the ID value to the request ID, which may break features that expect the "regular" ID. If you need to override the hiddenId for requests, maybe we need a new/different hidden field value.

Apologies if I'm missing/misunderstanding something -- but I suspect this bit needs a little more testing/review.

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
We're already using getLabelList above. Do we need it again here? Or should this supersede the earlier use?

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
The long check `$this->driver->supportsAjaxStatus() && $this->driver->hasCatalog() && $this->driver->pubTypeRtacEnabled()` is repeated several times in this template. I wonder if it's worth setting a variable at the top, like `$supportsRtac = $this->driver->supportsAjaxStatus() && $this->driver->hasCatalog() && $this->driver->pubTypeRtacEnabled()` so the later code can be simplified. Might slightly improve readability.

demiankatz reviewed (changes requested):
Thanks, @cwolfebsco, this is coming together very nicely. See below for a few comments, mostly minor -- though I think there may be a bit more significant work needed in the template.

cwolfebsco committed:
Update module/VuFind/src/VuFind/Controller/HoldsTrait.php

thank you!

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
Update module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php

quickly tested at my end and it seems to work exactly as suggested by you. Lovely feature to learn :)

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
remove superfluous test

based on recommendation, removing this test

cwolfebsco committed:
whitespace desaster

cwolfebsco commented on module/VuFind/tests/unit-tests/src/VuFindTest/RecordTab/HoldingsILSTest.php:
Indeed, seems to make sense. Of course I am terrible embarrassed by having the same code in two tests :( Not sure what I was thinking - likely I wasn't thinking...

cwolfebsco committed:
introduce supportsRtac as a single variable

lengthy evaluation used three times - change suggested by Demian and he is (of course) right about it

cwolfebsco commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
that makes a ton of sense, and I implemented as suggested. Thanks!

cwolfebsco committed:
remove getLabelList()

used further up and doesn't seem to return anything in testing

cwolfebsco commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
I've removed it, must have borrowed from elsewhere without understanding what "getLabelList()" actually does. Sad to say I am still not sure what it does :(

cwolfebsco commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
Thank you for pointing this out, obviously a naive implementation. I am having a "play" with adding a new ID and then updating check_item_statuses.js to account for this - the goal has to be to avoid availability checks to be issued for things that aren't from a custom catalog database

demiankatz commented on module/VuFind/tests/unit-tests/src/VuFindTest/RecordTab/HoldingsILSTest.php:
I imagine you mapped out your test cases first, then filled them in... and two of them just coincidentally happened to end up being the same. :-)

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
`getLabelList()` displays arbitrary labels that can be placed on record drivers. At present, this mechanism is only used in one place: the "Blender" backend that combines records from multiple sources uses labels to indicate which backend is the source of each record in the result set.

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
On a related note: you've also added a call to `getFormatList` here. Not sure if that's intentional. It doesn't seem to duplicate any existing logic, but I wonder if the format list was previously intentionally excluded. I haven't had time to do side-by-side comparison yet to see if this causes any undesirable changes, but maybe we should double-check that.

cwolfebsco committed:
introducing hiddenRTACId

hiddenId is used for functions beyond availability checking.

EDS > result-list.phtml introduces a new value "hiddenRTACId". For EDS this will always be present and either have the value of the appropriate id or "no_rtac"

if hiddenRTACId is present, it will replace the hiddenIdEl for availability checks.

For any id equalling "no_rtac" and early exit in the function avoid superfluous availability checks.

this approach caused an inconvenience in "lists" a list-entry.phtml was added to RecordDriver/EDS/ to also introduce the hiddenRTACId

cwolfebsco commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
<img width="954" height="322" alt="Screenshot 2025-09-03 at 12 14 51" src="https://github.com/user-attachments/assets/89be366d-05d7-4ffc-8db6-3c10a91fa0f1" />

You make a good point, I had implemented it for the sake of consistency to how default records show. But i also see why it might have been left out (on purpose).

In this screenshot you will notice that there is a publication type icon on the left where usually there would be a book over (EDS.css + appropriate classes) and it does state the Publication Type below the icon, so the blue labels could be seen as a duplication.

I guess this is one we need to make a decision on....

cwolfebsco committed:
remove comment

left a to do comment after I had addressed it

cwolfebsco commented on module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Logic/HoldsTest.php:
This to do is no longer needed - thanks for catching it.

cwolfebsco commented:
An area to consider is what to do with 

MyResearch/Fines
MyResearch/CheckedOut
Holds/List

This areas default back to a plain text display without a link, if a record cannot be found in the SOLR index. one could, with relative ease make a change to the account-entry.phtml to have these link back to the API based record. Not sure this needs to be addressed - would appreciate opinions.

<img width="635" height="308" alt="Screenshot 2025-09-03 at 13 03 41" src="https://github.com/user-attachments/assets/09edb6f1-def7-440d-afdf-ec0830b13e5c" />

In this screenshot the first record with hyperlink also has a solr-version of the record, while the second only had a EDSAPI record.

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
The easiest decision is probably to omit it -- since the format list is not directly related to the functionality being added here, it could be considered an out-of-scope change. If we want to add it, we could always do so in a separate PR, but I think you're right that it was originally omitted on purpose.

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
The counter-argument, of course, is that the icon only shows one format, while the list can show multiples (as it does in your example). But I still think it's fine to defer the decision to a separate/later PR, to keep the scope of this one contained.

demiankatz commented:
> An area to consider is what to do with
> 
> MyResearch/Fines MyResearch/CheckedOut Holds/List
> 
> This areas default back to a plain text display without a link, if a record cannot be found in the SOLR index. one could, with relative ease make a change to the account-entry.phtml to have these link back to the API based record. Not sure this needs to be addressed - would appreciate opinions.

The solution to this problem is to have the ILS driver return an appropriate 'source' value alongside the 'id' values in methods like getMyFines, etc. The source defaults to "Solr" but if the driver returned "EDS" instead, the correct record would load in this context.

Of course, this is a bit tricky -- if we wanted to implement this at the record driver level, we would have to make hard-to-test changes and add new configuration settings to every ILS driver, and there are a lot of them.

Perhaps a compromise would be to add a "default_record_backend" setting to the [Catalog] section of config.ini, and then adjust all of the relevant controller actions (for example, see this line: https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/Controller/MyResearchController.php#L1700) to fall back on the configured default value if available before further falling back on DEFAULT_SEARCH_BACKEND.

This way, somebody with an EDS backend can make one configuration change and all ILS drivers will be assumed to refer to EDS records instead of Solr records. It would require more work if we wanted different ILS drivers to connect to different backends -- I can see how that might be useful, but it's harder to implement and could be deferred until we have an actual use case.

Does that make any sense?

cwolfebsco commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
I will remove them, as this is not relevant to this functionality and I agree it can be handled in another PR, if we wanted that. Appreciate the eye for detail and the pragmatic approach...

cwolfebsco committed:
removing "getFromatList"

The format is shown below the publication type icon to the left of the record. While this is different to the rest of VuFind this change has no functional significance to the focus of the PR and I am thus removing it, as I had previously with the misunderstood getLabelList function call.

Thank you for pointing out this

demiankatz commented on themes/bootstrap5/js/check_item_statuses.js:
Would it be better to call this hiddenOverrideId for consistency with the naming used elsewhere in the code?

demiankatz commented on themes/bootstrap5/js/check_item_statuses.js:
Do we need this special value? I'm not sure of the circumstances in which this would be used, but just wondering if there's an opportunity to avoid a "magic value"?

demiankatz reviewed (changes requested):
Thanks, @cwolfebsco, see below for a couple of thoughts on the latest changes. (And I realize you probably also have some other work in progress at the moment -- just watching the changes as they come through :-) ).

cwolfebsco committed:
use hiddenOverrideId

rather than check for no_rtac check for empty

Reason for this hiddenIdEl.value empty:

In EDS API we will get mixed results between books (catalog) and articles. Articles by definition won't have availability from the ILS. As the hiddenId will always be present, we need a way to differentiate one from the other, to avoid overloading the ILS backend with superfluous requests

cwolfebsco commented on themes/bootstrap5/js/check_item_statuses.js:
I'll update this. As a matter of fact I tried to push that change this morning, but it is still processing. Sounds like GitHub needed an early weekend :)

cwolfebsco commented on themes/bootstrap5/js/check_item_statuses.js:
In the commit I pushed this morning it will simply evaluate it being empty. The key problem to avoid is for the check_item_statuses.js to fire on non catalog records. When I added the no_rtac I figured it was quite explicit, but I also see how it makes it harder in some way. Have a lovely weekend!

cwolfebsco committed:
Merge remote-tracking branch 'upstream/dev' into Patron-Empowerment-for-catalogs-in-EDS-API

demiankatz commented:
Thanks for the latest progress, @cwolfebsco -- your most recent changes look good to me. Are you still working through some of the other hidden ID implications? I notice that you haven't formally requested a new review yet -- just wondering if you need me to look more closely again, or if I should keep waiting. :-)

cwolfebsco commented:
@demiankatz - Thank you for following up - this is much appreciated. I've been pondering the implication on the MyResearch area a bit and believe this likely is an acceptable state, to be possibly re-visited, once a library takes up that option.

I wish to express my gratitude for your kindness and patience throughout the process. I fear you will uncover more items to be looked at and I will do my best to address them quickly. This project is lucky to have you

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/list-entry.phtml:
It may be possible to reduce the amount of duplication in this template. If you add a check to the default list-entry.phtml template that checks if `$this->overrideId` is set and generates the `hiddenOverrideId` element if so, then you might be able to reduce this template down to something like:

```
<?php
  $this->assetManager()->appendStyleLink('EDS.css');
  $supportsRtac = $this->driver->supportsAjaxStatus() && $this->driver->hasCatalog() && $this->driver->pubTypeRtacEnabled();
  $this->overrideId = $supportsRtac ? $this->driver->getUniqueIDOverrideForRequest() : '';
  $this->render('RecordDriver/DefaultRecord/list-entry.phtml');
```

...I apologize if I'm misremembering some detail about scoping that makes that not work as I think it will (in which case, I'm sure there IS a way to make it work, the syntax might just be slightly different). But I think I'm probably on target here, and doing this will make the code a lot easier to maintain in future.

What do you think?

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
These lines are outside of the if/else, so they don't need to be indented:

```suggestion
<input type="hidden" value="<?=$this->escapeHtmlAttr($this->driver->getUniqueID())?>" class="hiddenId">
<input type="hidden" value="<?=$this->escapeHtmlAttr($this->driver->getSourceIdentifier())?>" class="hiddenSource">
```

demiankatz commented on themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml:
Since there are no formats in here, maybe we should take off the result-formats class; not sure why it's there, but maybe just a copy-paste from earlier.

demiankatz reviewed (changes requested):
Thanks, @cwolfebsco, I gave this a closer review and have a few more suggestions.

Also, if you think it would be helpful to add the "default backend for ILS drivers" configuration I suggested earlier, I might be able to put that together quickly in a separate PR. But as you say, maybe it's just as well to wait until we have a known user before investing time/effort.

cwolfebsco committed:
removing enclosing div.result-formats

removing enclosing div-results-formats - this was down to a misunderstanding at my end, when borrowing from the main template

cwolfebsco committed:
Update themes/bootstrap5/templates/RecordDriver/EDS/result-list.phtml

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

cwolfebsco committed:
remove duplication of code

DefaultRecord/list-entry.phtml -> checks for existence of overrideId to add hiddenOverrideId in
EDS/list-entry.phtml injects the CSS, sets overrideId and then echos the render of the DefaultRecord/list-entry.phtml template

cwolfebsco committed:
empty line at end

cwolfebsco committed:
Merge remote-tracking branch 'upstream/dev' into Patron-Empowerment-for-catalogs-in-EDS-API

cwolfebsco commented on themes/bootstrap5/templates/RecordDriver/EDS/list-entry.phtml:
Much nicer - my commit somehow didn't process - something for me to look at tomorrow, I fear

cwolfebsco committed:
Merge remote-tracking branch 'upstream/dev' into Patron-Empowerment-for-catalogs-in-EDS-API

cwolfebsco committed:
adding static validTitle, validAuthor, validPublisher

during conflict resolution, I screwed this one up :(

cwolfebsco committed:
add comments

add comments to hopefully create a bit of clarity

cwolfebsco committed:
no trailing whitespace

cwolfebsco committed:
Merge remote-tracking branch 'upstream/dev' into Patron-Empowerment-for-catalogs-in-EDS-API

demiankatz commented on themes/bootstrap5/templates/RecordDriver/DefaultRecord/list-entry.phtml:
Maybe better to use a PHP comment here so the comment doesn't increase the size of the rendered output sent to the user:

```suggestion
  <?php if (isset($this->overrideId)): // hiddenOverrideId is used for availability checking for non SOLR records?>
```

demiankatz reviewed (changes requested):
I think we're just about done, @cwolfebsco -- just one last very minor suggestion.

cwolfebsco committed:
Update themes/bootstrap5/templates/RecordDriver/DefaultRecord/list-entry.phtml

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

demiankatz commented:
@cwolfebsco, before I do the final round of testing on this, one more question: do we need to support title-level holds? I imagine some work on \VuFind\ILS\Logic\TitleHolds is needed if so.

cwolfebsco commented:
> @cwolfebsco, before I do the final round of testing on this, one more question: do we need to support title-level holds? I imagine some work on \VuFind\ILS\Logic\TitleHolds is needed if so.

yes, I fear I completely ignored title holds. Let me work on that next week. Thanks for the continued guidance

cwolfebsco commented:
@demiankatz 

you might need to educate me a little bit about Title Level Holds, please. 

From what I gather title level holds will be used if level === title (e.g. https://github.com/vufind-org/vufind/blob/ab7c0d4cffee33bc9ea76426f3838d2de2c0cf04/module/VuFind/src/VuFind/ILS/Driver/Folio.php#L2152) and then it changes the baseParams (https://github.com/vufind-org/vufind/blob/ab7c0d4cffee33bc9ea76426f3838d2de2c0cf04/module/VuFind/src/VuFind/ILS/Driver/Folio.php#L2153)

Now while I was poking at this I found that the level was always null as it is never set here: https://github.com/vufind-org/vufind/blob/ab7c0d4cffee33bc9ea76426f3838d2de2c0cf04/module/VuFind/src/VuFind/ILS/Driver/Folio.php#L928

But that might be a bit of a red herring, when I forced it to be level = 'title', the current functionality works as set.

I conclude that I have - as so often - too little understanding and would appreciate some guidance

cwolfebsco commented:
@demiankatz but then again there might be more that I need to look at. When I compared to a solr record - I see the title level request
<img width="984" height="436" alt="Screenshot 2025-10-03 at 15 51 04" src="https://github.com/user-attachments/assets/7b9a3ea5-c2cd-4c49-8718-f4e07ef763d6" />

Just in EDS it doesn't show:
<img width="1065" height="436" alt="Screenshot 2025-10-03 at 15 51 50" src="https://github.com/user-attachments/assets/82b2b093-3bd7-40e4-87bc-c316b40b46ec" />

In other words I have a different riddle to figure out :)

demiankatz commented:
@cwolfebsco, the hold level of 'title' should be set by the hold form when title level holds are active/displayed.

What setting do you have for title-level holds (`title_level_holds_mode`) in your config.ini? It might be interesting to see if you get different results for `always` vs. `driver`. That may narrow down whether the problem is in the ILS driver or the template/controller.

I confess that it's been a while since I've refreshed my own memory on how this all works, but if you need more guidance, I'll be happy to do that (probably early next week).

cwolfebsco committed:
Support Title Level Holds

Support Title Level Holds in EDS API

cwolfebsco committed:
Merge remote-tracking branch 'upstream/dev' into Patron-Empowerment-for-catalogs-in-EDS-API

cwolfebsco committed:
doc comment fixes

cwolfebsco committed:
added LinkOverrides

driverHold called on getHoldDetails without linkOverrides - fixed now

cwolfebsco committed:
fixing doc comments (spaces)

cwolfebsco committed:
Merge remote-tracking branch 'upstream/dev' into Patron-Empowerment-for-catalogs-in-EDS-API

cwolfebsco commented:
Title level requests should now be supported

demiankatz reviewed (approved):
Thanks, @cwolfebsco, this is looking good to me now, and all tests are passing -- I'm merging now!
